src/cloud-api-adaptor/pkg/securecomms/ppssh/ppssh.go

```diff
@@ -0,0 +1,304 @@
+	}
+
+	go func() {
+		kubernetesPhaseConfig := attestationPhase(listener, inbounds, outbounds, NewPpSecrets(getSecret))
+		for {
+			k8sPhase(listener, inbounds, outbounds, kubernetesPhaseConfig)
+		}
+	}()
+}
+
+func Singleton() {
+	// Signleton- make sure we run the ssh service once per boot.
+	if _, err := os.Stat(SIGNELTON_PATH); !errors.Is(err, os.ErrNotExist) {
+		log.Fatal("SSH Service run in signleton mode - cant be executed twice")
+	}
+	singleton, err := os.Create(SIGNELTON_PATH)
+	if err != nil {
+		log.Fatal(err)
+	}
+	singleton.Close()
+}
+
+func getAttestationPhaseKeys() (ppPrivateKeyBytes []byte, tePublicKeyBytes []byte) {
+	var err error
+
+	// Attestation Phase - may have unproven tePublicKeyBytes
+	tePublicKeyBytes, err = os.ReadFile(UNPROVEN_WN_PUBLIC_KEY_PATH)
+	if err != nil {
+		tePublicKeyBytes = nil
+	}
+
+	// Private Key generation - unproven to the clinet, key is generated on the fly
+	ppPrivateKey, err := rsa.GenerateKey(rand.Reader, 3072)
+	if err != nil {
+		log.Fatalf("Attastation Phase: Failed to generate host key, err: %v", err)
+	}
+
+	// Validate Private Key
+	err = ppPrivateKey.Validate()
+	if err != nil {
+		log.Fatalf("Attastation Phase: Failed to validate host key, err: %v", err)
+	}
+
+	ppPrivateKeyBytes = sshutil.RsaPrivateKeyPEM(ppPrivateKey)
+	log.Printf("Attastation Phase: SSH Server initialized keys")
+	return
+}
+
+func setConfigHostKey(config *ssh.ServerConfig, ppPrivateKeyBytes []byte) error {
+	serverSigner, err := ssh.ParsePrivateKey(ppPrivateKeyBytes)
+	if err != nil {
+		return fmt.Errorf("ssh.ParsePrivateKey of ppPrivateKeyBytes %w", err)
+	}
+	config.AddHostKey(serverSigner)
+	return nil
+}
+
+func setPublicKey(config *ssh.ServerConfig, tePublicKeyBytes []byte) error {
+	teSshPublicKey, _, _, _, err := ssh.ParseAuthorizedKey(tePublicKeyBytes)
+	if err != nil {
+		return fmt.Errorf("ssh.ParseAuthorizedKey of tePublicKeyBytes %w", err)
+	}
+
+	// An SSH server is represented by a ServerConfig, which holds
+	// certificate details and handles authentication of ServerConns.
+	config.PublicKeyCallback = func(c ssh.ConnMetadata, clientPublicKey ssh.PublicKey) (*ssh.Permissions, error) {
+		if bytes.Equal(teSshPublicKey.Marshal(), clientPublicKey.Marshal()) {
+			return &ssh.Permissions{
+				// Record the public key used for authentication.
+				Extensions: map[string]string{
+					"pubkey-fp": ssh.FingerprintSHA256(clientPublicKey),
+				},
+			}, nil
+		}
+		return nil, fmt.Errorf("unknown public key for %q", c.User())
+	}
+	return nil
+}
+
+func InitAttestationPhaseSshConfig() (*ssh.ServerConfig, error) {
+	config := &ssh.ServerConfig{}
+
+	ppPrivateKeyBytes, tePublicKeyBytes := getAttestationPhaseKeys()
+
+	if tePublicKeyBytes != nil { // connect with an client public key
+		if err := setPublicKey(config, tePublicKeyBytes); err != nil {
+			return nil, err
+		}
+	} else {
+		config.NoClientAuth = true
+		log.Printf("Attastation Phase: SSH Server initialized with NoClientAuth")
```

Comment:
Shouldn't that be elevated to a warning?

Reply:
No warning unfortunately in the currently used log package
I would prefer to leave it to another PR to replace the log package used at the entire cloud-api-adaptor